### PR TITLE
Added in core/utils documentation

### PR DESCRIPTION
**Pull request by iquah1**
1) Added in documentation for core/utils documentation
2) added in changes to the main `rst` 
3) handled a file encoding issue

### CODE REVIEW

File: src/pke/include/scheme/cryptocontextparams-base.h [hunk 1]
```diff
@@ -56,7 +56,7 @@ namespace lbcrypto {
 TODO (dsuponit): Do we need to keep these pairs cyclOrder / ringDimention and numPrimes / multDepth in Params?
     We should remove one variable from each pair.Otherwise we will complicate the code by comparing these parameters and
     making a decision for the user who sets both of them at the same time.it is going to be very ambiguousand will
-    lead us to confusions and bugs.Either way, we must stay with just ONE set of these parameters.It doesn�t really matter
+    lead us to confusions and bugs.Either way, we must stay with just ONE set of these parameters.It doesn�t really matter
```
Comment (iquah1) [inline src/pke/include/scheme/cryptocontextparams-base.h:59]: The file encoding here was ...? Anyways, it was causing sphinx to run into errors

File: src/core/include/utils/README.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Misc. Utils
+
+Contains various utilities across:
+
+- debugging utilities
+- memory management utilities
+- prallel operations
```
Comment (mtriplett-duality) [inline src/core/include/utils/README.md:7]: prallel -> parallel

File: src/pke/include/scheme/cryptocontextparams-base.h (hunk 1 above)
Comment (mtriplett-duality) [reply src/pke/include/scheme/cryptocontextparams-base.h:59]: Got it, makes sense.

**Comment by dsuponitskiy** [comment]
The branch doesn't seem to have been rebased

**Comment by iquah1** [comment]
@mtriplett-duality @dsuponitskiy 

This has been fully rebased and is ready for a second round of reviews. No rush on this since I'm going to be on the other project next week and I don't think we should merge stuff in Friday at 3:30

File: docs/assets/sphinx_rsts/modules/core/lattice/hal.rst [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+Core Lattice Hardware Abstraction Layer Documentation
+======================================================
+
+.. mermaid::
+
+  graph BT
+      a[DCRTPolyInterface] --> |Inherited by|b[DCRTPoly - HAL Default];
+      b[DCRTPoly-HAL Default] --> |Inherited by|c[HexlDCRTPoly];
+
+This hardware abstraction layer allows OpenFHE to use a variety of device backends while still allowing for high performance. As of March 31st 2022, we have the default backend, and the `Intel Hexl <https://github.com/intel/hexl>`_ backend.
```
Comment (mtriplett-duality) [inline docs/assets/sphinx_rsts/modules/core/lattice/hal.rst:12]: While I'm not sure we want to specify hardware specifically, I feel hardware-integrated backends may be clearer towards that point here.

File: docs/assets/sphinx_rsts/modules/core/lattice/hal.rst [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+   :backlinks: none
+
+
+Core Lattice HAL Documentation
+-------------------------------
+
+.. autodoxygenindex::
+   :project: core_lattice_hal
+
+Core Lattice HAL HEXL Documentation
```
[the remark this answers is not in the document]
Comment (mtriplett-duality) [reply docs/assets/sphinx_rsts/modules/core/lattice/hal.rst:28]: Nevermind! Got it.

**Review by mtriplett-duality — APPROVED** [review]

**Pull request merged**